### PR TITLE
Add showkase wrapper tool

### PR DESCRIPTION
**Pull request by AsimRibo**
### Type

- [x] **Feature**: This pull request introduces a new feature.
- [ ] **Bug fix**: This pull request fixes a bug.
- [ ] **Refactor**: This pull request refactors existing code.
- [ ] **Documentation**: This pull request updates documentation.
- [ ] **Other**: This pull request makes other changes.

#### Additional information

- [ ] This pull request introduces a **breaking change**.

### Description

Since Showkase is now often used in projects for displaying existing composables, I investigated possibility of adding it to Sentinel to avoid having to create a custom Sentinel tool in our projects. The resulting setup should make it easier to run Showkase inside of Sentinel.

Still have some testing to do.

## Checklist

- [x] I have performed a self-review of my own code.
- [x] I have tested my changes, including edge cases.
- [ ] I have added necessary tests for the changes introduced (if applicable).
- [ ] I have updated the documentation to reflect my changes (if applicable).

### CODE REVIEW

File: tool-showkase/consumer-rules.pro [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+# The generated class is annotated with @ShowkaseRootCodegen, which has RUNTIME retention.
+-keepattributes RuntimeVisibleAnnotations
+
+# The app's @ShowkaseRoot class, whose canonical name ShowkaseTool passes to the browser.
+-keep class * implements com.airbnb.android.showkase.annotation.ShowkaseRootModule { *; }
+
+# The generated provider, matched two ways in case R8 cannot see one of the supertypes.
+# Members are kept too because the class is only ever instantiated reflectively.
+-keep class * implements com.airbnb.android.showkase.models.ShowkaseProvider { *; }
+-keep @com.airbnb.android.showkase.annotation.ShowkaseRootCodegen class * { *; }
```
Comment (KCeh) [inline tool-showkase/consumer-rules.pro]: Is this correct? I think there is a chance for R8 to fail (missing class)

Let's double-check this

File: tool-showkase/src/main/kotlin/com/infinum/sentinel/ui/tools/ShowkaseTool.kt [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+    override fun listener(): View.OnClickListener = View.OnClickListener { view ->
+        try {
+            view.context.startActivity(
+                ShowkaseBrowserActivity.getIntent(view.context, rootModule.java.canonicalName.orEmpty()).apply {
+                    flags = Intent.FLAG_ACTIVITY_SINGLE_TOP
+                    addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
+                },
+            )
+        } catch (error: NoClassDefFoundError) {
```
Comment (KCeh) [inline tool-showkase/src/main/kotlin/com/infinum/sentinel/ui/tools/ShowkaseTool.kt:39]: What about ActivityNotFoundException?

File: tool-showkase/proguard-rules.txt [hunk 3]
```diff
@@ -0,0 +1,4 @@
+-keeppackagenames
+-keep public class com.infinum.sentinel.ui.tools.ShowkaseTool {
+  public protected *;
+}
```
Comment (KCeh) [inline tool-showkase/proguard-rules.txt:4]: I suggest following other proguard configs and add:
-keep public class com.infinum.sentinel.databinding.**

just in case

File: tool-showkase/README.md [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+The tool is only a launcher. Showkase's annotation processor has to run over your own sources, so
+the Showkase build setup stays in your app.
+
+## Setup
+
+**`app/build.gradle`**
+
+```groovy
+def sentinelVersion = "2.0.0"
```
Comment (KCeh) [inline tool-showkase/README.md]: I suggests ether omitting concrete versions or using an automated approach (like in replaceVersionsInFile) for keeping them up to date. We will forget to update versions here sooner or later

**Review by KCeh — COMMENTED** [review]
Looks solid 🚀 

We still need to add tool to sample app and then we can do testing 👍 

Also, small note for upcoming release: we should probably try to resolve (at least some of) compile warnings

File: tool-showkase/src/main/kotlin/com/infinum/sentinel/ui/tools/ShowkaseTool.kt (hunk 2 above)
Comment (AsimRibo) [reply tool-showkase/src/main/kotlin/com/infinum/sentinel/ui/tools/ShowkaseTool.kt:39]: I don't think it can happen since activity is always declared in showkase's Manifest so you will always have either both the activity and declaration or none. Don't think anything else can produce ActivityNotFoundException.

Reason why I have put NoClassDefFoundError is because only way this can fail is if client project did not include showkase dependency.

File: tool-showkase/consumer-rules.pro (hunk 1 above)
Comment (AsimRibo) [reply tool-showkase/consumer-rules.pro]: Not much experience with this so I was winging it a bit.  I have tried deploying Sentinel locally and including it in client project, it worked with rules above. I did reconsider this part and adjusted the rules which I will push later. There is section in showkase README that says rules are not even needed, but I didn't find that to be the case when using it.

File: tool-showkase/proguard-rules.txt (hunk 3 above)
Comment (AsimRibo) [reply tool-showkase/proguard-rules.txt:4]: Will do.

File: tool-showkase/README.md (hunk 4 above)
Comment (AsimRibo) [reply tool-showkase/README.md]: Omitting the versions sounds simpler ngl 😄 .

File: tool-showkase/README.md [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+dependencies {
+    debugImplementation   "com.infinum.sentinel:tool-showkase:$sentinelVersion"
+    releaseImplementation "com.infinum.sentinel:tool-showkase-no-op:$sentinelVersion"
+
+    // The browser runtime. It carries the whole Compose stack, so keep it out of release.
+    debugImplementation   "com.airbnb.android:showkase:$showkaseVersion"
+
+    // The processor. Use a variant-scoped configuration, never the project-wide `ksp`.
+    kspDebug              "com.airbnb.android:showkase-processor:$showkaseVersion"
```
[the remark this answers is not in the document]
Comment (KCeh) [reply tool-showkase/README.md]: This is maybe worth mentioning in tool readme

File: tool-showkase/src/main/kotlin/com/infinum/sentinel/ui/tools/ShowkaseTool.kt (hunk 2 above)
Comment (KCeh) [inline tool-showkase/src/main/kotlin/com/infinum/sentinel/ui/tools/ShowkaseTool.kt:39]: I saw one project that implemented custom Showkase tool in simpler way (no rootModule was needed):
`context.startActivity(Showkase.getBrowserIntent(context).addFlags(Intent.FLAG_ACTIVITY_NEW_TASK))`

Where Showkase is

```
package com.airbnb.android.showkase.models

/**
 * Object that will be the receiver for all extension functions that are generated as part of
 * Showkase. Some examples of extension funtions that will be generated for your use are
 * "createShowkaseBrowserIntent" that allows you to start the ShowkaseBrowser and
 * "getShowkaseMetadata" that gives you access to all the metadata that's available to Showkase.
 */
object Showkase
```

can we make API simpler? Or was that some project-specific trick?

File: gradle/libs.versions.toml [hunk 6]
```diff
@@ -53,6 +55,8 @@ toolgoogleplay = { module = "com.infinum.sentinel:tool-googleplay", version.ref
 tooltimber = { module = "com.infinum.sentinel:tool-timber", version.ref = "sentinel" }
 toolnetworkemulatorokhttp = { module = "com.infinum.sentinel:tool-networkemulator-okhttp", version.ref = "sentinel" }
 toolnetworkemulatorokhttpnoop = { module = "com.infinum.sentinel:tool-networkemulator-okhttp-no-op", version.ref = "sentinel" }
+toolshowkase = { module = "com.infinum.sentinel:tool-showkase", version.ref = "sentinel" }
+toolshowkasenoop = { module = "com.infinum.sentinel:tool-showkase-no-op", version.ref = "sentinel" }
```
Comment (KCeh) [inline gradle/libs.versions.toml:59]: Can we add those to bundles section of version catalog? so we don't forget

**Review by KCeh — COMMENTED** [review]
Looks good, before I hit approve can you just quickly check one thing:
Can we simplify tool API - check my comment for more info

File: tool-showkase/src/main/kotlin/com/infinum/sentinel/ui/tools/ShowkaseTool.kt (hunk 2 above)
Comment (AsimRibo) [reply tool-showkase/src/main/kotlin/com/infinum/sentinel/ui/tools/ShowkaseTool.kt:39]: It is similarly implemented on my project as well, but I would say it is project specific. Down below you can see `Showkase.getBrowserIntent` function which is autogenerated, not part of Showkase runitime. 
```
/**
 * Helper function that's autogenerated and gives you an intent to start the ShowkaseBrowser.
 */
public fun Showkase.getBrowserIntent(context: Context): Intent {
    val intent = Intent(context, ShowkaseBrowserActivity::class.java)
    intent.putExtra("SHOWKASE_ROOT_MODULE", "some.package.MyRootModule")
    return intent
}
```

The processor generates it into the app, with the root module name hardcoded as a string. A library can't reference app-generated code, so the app has to hand us the root module, which is what the `KClass` parameter does. Under the hood it is the same intent our tool builds via `ShowkaseBrowserActivity.getIntent`. That project could use the helper only because their tool class lived in the app module itself.

File: gradle/libs.versions.toml (hunk 6 above)
Comment (AsimRibo) [reply gradle/libs.versions.toml:59]: Added `toolshowkase` without no-op, as per convention. Maybe we could also bundle no-op:
```
toolsnoop = ["toolnetworkemulatorokhttpnoop", "toolshowkasenoop"]
```
but since there are only 2 present right now I would say no need.

File: sample/build.gradle.kts [hunk 7]
```diff
@@ -7,6 +7,8 @@ val releaseConfig = extra["releaseConfig"] as Map<String, Any>
 plugins {
     id("com.android.application")
     id("org.jetbrains.kotlin.android")
+    alias(libs.plugins.kotlin.compose)
```
[the remark this answers is not in the document]
Comment (AsimRibo) [reply sample/build.gradle.kts:10]: Maybe I am wrong, but I think this was the case for previous AGP versions. Compose plugin should be the switch now. Regardless, it is working 🤷 .
@KCeh

**Review by KCeh — APPROVED** [review]
Ok looks good 🚀 

I will asssist with checking lib before release
